### PR TITLE
Fixed #34427 -- Improved error message when context processor does not return a dict.

### PR DESCRIPTION
**Pull request by shangxiao**

### CODE REVIEW

File: django/template/context.py [hunk 1]
```diff
@@ -251,7 +251,14 @@ def bind_template(self, template):
         processors = template.engine.template_context_processors + self._processors
         updates = {}
         for processor in processors:
-            updates.update(processor(self.request))
+            try:
+                updates.update(processor(self.request))
+            except TypeError as e:
+                raise TypeError(
+                    "Context processor %s didn't return a dictionary."
+                    % processor.__name__
```
Comment (felixxm) [inline django/template/context.py:259]: I'd use f-string and `__qualname__`:
```suggestion
                    f"Context processor {processor.__qualname__} didn't return a dictionary."
```

File: tests/template_tests/test_context.py [hunk 2]
```diff
@@ -276,3 +280,27 @@ def test_modify_context_and_render(self):
         context = RequestContext(request, {})
         context["foo"] = "foo"
         self.assertEqual(template.render(context), "foo")
+
+    @override_settings(
+        TEMPLATES=[
+            {
+                "BACKEND": "django.template.backends.django.DjangoTemplates",
+                "APP_DIRS": True,
```
Comment (felixxm) [inline tests/template_tests/test_context.py:288]: This is unnecessary
```suggestion
```

File: tests/template_tests/test_context.py [hunk 3]
```diff
@@ -276,3 +280,27 @@ def test_modify_context_and_render(self):
...
+    )
+    def test_template_context_processor_returning_none(self):
+        request_context = RequestContext(HttpRequest())
+        msg = (
+            "Context processor context_process_returning_none didn't return a "
+            "dictionary."
+        )
+        with self.assertRaisesMessage(TypeError, msg):
+            with request_context.bind_template(Template("")):
+                ...
```
[the remark this answers is not in the document]
Comment (shangxiao) [reply tests/template_tests/test_context.py:306]: Ah yeah I should know that 😅

File: django/template/context.py [hunk 4]
```diff
@@ -251,7 +251,14 @@ def bind_template(self, template):
         processors = template.engine.template_context_processors + self._processors
         updates = {}
         for processor in processors:
-            updates.update(processor(self.request))
+            try:
+                updates.update(processor(self.request))
```
Comment (timgraham) [inline django/template/context.py:255]: What about moving `processor(self.request)` outside the try/except so the exception isn't wrapped if `processor(self.request)` raises `TypeError`?

Comment (shangxiao) [reply django/template/context.py:255]: Good point

File: django/template/context.py (hunk 1 above)
Comment (shangxiao) [reply django/template/context.py]: Actually just noticed the extra `f` on this, have removed it.

File: django/template/context.py [hunk 5]
```diff
@@ -251,7 +251,15 @@ def bind_template(self, template):
...
         for processor in processors:
-            updates.update(processor(self.request))
+            context = processor(self.request)
+            try:
+                updates.update(context)
+            except TypeError as e:
+                raise TypeError(
+                    f"Context processor {processor.__qualname__} didn't return a "
+                    "dictionary."
+                ) from e
```
Comment (adamchainz) [inline django/template/context.py:261]: `as e` / `from e` are unnecessary, since that’s the default behaviour. But oh well.

**Review by adamchainz — COMMENTED** [review]
Speedy work, nice one 👍

File: django/template/context.py (hunk 5 above)
Comment (shangxiao) [reply django/template/context.py:261]: @adamchainz You're probably more familiar with chaining than me but isn't the implicit version slightly different? 🤔  At least that's how I'm reading it in the docs:

https://docs.python.org/3/tutorial/errors.html#exception-chaining

More specifically: https://docs.python.org/3/reference/simple_stmts.html#raise

Using `from` attaches the parent to the `__cause__` as opposed to the `__context__` with the implicit variety 🤔

Comment (timgraham) [reply django/template/context.py:261]: I checked this too. Without `from e` the message that's printed says, "During handling of the above exception, another exception occurred:" but with it "This exception is the direct cause of the following exception".

Comment (adamchainz) [reply django/template/context.py:261]: Right y'all are. My bad!

**Pull request merged**